Реализована функциональность по ТЗ №7

OlegDyuzhin opened the pull request:
Добавлена функциональность таскменеджеру автосохранения в файл и восстановления из файла

OlegDyuzhin committed:
Реализована функциональность по ТЗ №7

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,148 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+
+    Path path = Path.of("src\\resources\\savedTasks.csv"); //файл по умолчанию
```
Для соблюдения инкапсуляции переменную лучше сделать приватной.

OlegDyuzhin replied:
Исправил

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,148 @@
+        if (path.toFile().exists()) {
+            this.path = path;
+        } else {
+            try {
```
Вообще говоря, не рекомендуется в конструкторах использовать try
Если в процессе выполнения возникло исключение, то может так получиться, что объект оказался не до конца сформированным, что может привести к дополнительным сложностям при отладке.  

На усмотрение.

OlegDyuzhin replied:
убрал try и проверки, те же проверки делает BufferedReader и BufferedWriter при создании через Files

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,148 @@
+                ep.setSubtasks(sub);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Здесь также лучше пробросить собственное исключение: 

} catch (IOException e) {
                throw new ManagerSaveException("Не удалось восстановить данные из файла");
            }

OlegDyuzhin replied:
исправил

avfyodorov reviewed (commented):
Добрый день, Олег! 

Очень хорошая работа, от меня всего пара небольших уточнений.

OlegDyuzhin committed:
Устранены замечания:
-убраны блоки try из конструкторов (проверки проходят в методах save и loadFromFile
-сделан небольшой рефакторинг

OlegDyuzhin committed:
-сделан небольшой рефакторинг в тестах

avfyodorov reviewed (commented):
Добрый вечер, Олег! 

Всё в порядке, замечаний нет. 
Работа принята.
